precise/iaf_psc_exp_ps_lossless.cpp

```diff
@@ -0,0 +1,671 @@
+/*
+ *  iaf_psc_exp_ps_lossless.cpp
+ *
+ *  This file is part of NEST.
+ *
+ *  Copyright (C) 2004 The NEST Initiative
+ *
+ *  NEST is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "iaf_psc_exp_ps_lossless.h"
+
+// C++ includes:
+#include <limits>
+
+// Includes from nestkernel:
+#include "exceptions.h"
+#include "universal_data_logger_impl.h"
+
+// Includes from sli:
+#include "dict.h"
+#include "dictutils.h"
+#include "integerdatum.h"
+#include "doubledatum.h"
+#include "arraydatum.h"
+
+
+/* ----------------------------------------------------------------
+ * Recordables map
+ * ---------------------------------------------------------------- */
+
+nest::RecordablesMap< nest::iaf_psc_exp_ps_lossless >
+  nest::iaf_psc_exp_ps_lossless::recordablesMap_;
+
+namespace nest
+{
+// Override the create() method with one call to RecordablesMap::insert_()
+// for each quantity to be recorded.
+template <>
+void
+RecordablesMap< iaf_psc_exp_ps_lossless >::create()
+{
+  // use standard names whereever you can for consistency!
+  insert_( names::V_m, &iaf_psc_exp_ps_lossless::get_V_m_ );
+  insert_( names::I_syn, &iaf_psc_exp_ps_lossless::get_I_syn_ );
+  insert_( names::I_syn_ex, &iaf_psc_exp_ps_lossless::get_I_syn_ex_ );
+  insert_( names::I_syn_in, &iaf_psc_exp_ps_lossless::get_I_syn_in_ );
+}
+}
```

Comment:
We need `namespace nest { ... }` here at the beginning for the *declaration* of the `RecordablesMap< iaf_psc_exp_ps_lossless >::create()` specialisation (which also is a definition); this has to happen in `namespace nest { ... }`.  All else are just definitions of methods that have already been declared in the header file, so we do not need the namespace around. Generally, we want to have as little as possible inside `namespace xyz { }` to avoid that we accidentally introduce names into a namespace.